ssi-sd-jwt/src/encode.rs

```diff
@@ -0,0 +1,218 @@
+
+    let salt = Base64UrlUnpadded::encode_string(&salt_bytes);
+
+    encode_disclosure_with_salt(&salt, claim_name, claim_value)
+}
+
+fn encode_disclosure<ClaimValue: Serialize>(
+    claim_name: Option<&str>,
+    claim_value: &ClaimValue,
+) -> Result<String, serde_json::Error> {
+    let mut rng = rand::rngs::OsRng {};
+    encode_disclosure_with_rng(&mut rng, claim_name, claim_value)
+}
+
+pub fn encode_property_disclosure<ClaimValue: Serialize>(
+    sd_alg: SdAlg,
+    claim_name: &str,
+    claim_value: &ClaimValue,
+) -> Result<Disclosure, serde_json::Error> {
+    let encoded = encode_disclosure(Some(claim_name), claim_value)?;
+    let hash = hash_encoded_disclosure(sd_alg, &encoded);
+
+    Ok(Disclosure { encoded, hash })
+}
+
+pub fn encode_array_disclosure<ClaimValue: Serialize>(
+    sd_alg: SdAlg,
+    claim_value: &ClaimValue,
+) -> Result<Disclosure, serde_json::Error> {
+    let encoded = encode_disclosure(None, claim_value)?;
+    let hash = hash_encoded_disclosure(sd_alg, &encoded);
+
+    Ok(Disclosure { encoded, hash })
+}
+
+pub fn encode_sign<Claims: Serialize>(
+    algorithm: Algorithm,
+    base_claims: &Claims,
+    key: &JWK,
+    sd_alg: SdAlg,
+    disclosures: Vec<UnencodedDisclosure>,
+) -> Result<(String, Vec<FullDisclosure>), EncodeError> {
+    let mut base_claims_json = serde_json::to_value(base_claims)?;
+
+    let post_encoded_disclosures: Result<Vec<_>, EncodeError> = disclosures
+        .iter()
+        .map(|disclosure| {
+            let encoded = disclosure.encode()?;
+            let hash = hash_encoded_disclosure(sd_alg, &encoded);
+            Ok(FullDisclosure {
+                encoded,
+                hash,
+                unencoded: disclosure.clone(),
+            })
+        })
+        .collect();
+
+    let post_encoded_disclosures = post_encoded_disclosures?;
+
+    {
+        let base_claims_obj = base_claims_json
+            .as_object_mut()
+            .ok_or(EncodeError::EncodedAsNonObject)?;
+
+        let prev_sd_alg = base_claims_obj.insert(
+            SD_ALG_CLAIM_NAME.to_owned(),
+            serde_json::json!(sd_alg.to_str()),
+        );
+
+        if prev_sd_alg.is_some() {
+            return Err(EncodeError::EncodedClaimsContainsReservedProperty);
+        }
+
+        let mut sd_claim = vec![];
+
+        for disclosure in post_encoded_disclosures.iter() {
+            match disclosure.unencoded {
+                UnencodedDisclosure::Property(ref claim_name, _) => {
+                    sd_claim.push(serde_json::Value::String(disclosure.hash.clone()));
+                    base_claims_obj.remove(claim_name);
+                }
+                UnencodedDisclosure::ArrayItem(ref claim_name, _) => {
+                    if !base_claims_obj.contains_key(claim_name) {
+                        let _ = base_claims_obj.insert(claim_name.clone(), serde_json::json!([]));
+                    }
+
+                    let array = base_claims_obj.get_mut(claim_name).unwrap();
```

Comment:
lingering unwrap

Reply:
The unwrap is because claim_name is explicitly added in all cases in the couple lines above, but this absolutely needed a comment justifying this.  That has been added.